proxy/src/main/java/com/velocitypowered/proxy/connection/client/HandshakeSessionHandler.java

```diff
@@ -85,31 +86,57 @@ public boolean handle(LegacyHandshake packet) {
   public boolean handle(Handshake handshake) {
     InitialInboundConnection ic = new InitialInboundConnection(connection,
         cleanVhost(handshake.getServerAddress()), handshake);
-    StateRegistry nextState = getStateForProtocol(handshake.getNextStatus());
-    if (nextState == null) {
-      LOGGER.error("{} provided invalid protocol {}", ic, handshake.getNextStatus());
-      connection.close(true);
-    } else {
-      connection.setProtocolVersion(handshake.getProtocolVersion());
-      connection.setAssociation(ic);
-
-      switch (nextState) {
-        case STATUS:
-          connection.setActiveSessionHandler(StateRegistry.STATUS,
-              new StatusSessionHandler(server, ic));
-          break;
-        case LOGIN:
-          this.handleLogin(handshake, ic);
-          break;
-        default:
-          // If you get this, it's a bug in Velocity.
-          throw new AssertionError("getStateForProtocol provided invalid state!");
-      }
-    }
+
+    // Handle connection establish event.
+    connection.setAutoReading(false);
+    server.getEventManager()
+            .fire(new ConnectionEstablishEvent(
+                    ic, getIntentionForStatus(handshake.getNextStatus())))
```

Comment:
If `getIntentionForStatus` returns null, I would prefer if we just closed the connection right then and there. But I can understand why we might want to check for an invalid status.